File: doc/porting/BalloonItem.cs
using Assets.Scripts.GamePlayScene.Mechanics.Animations;
using Assets.Scripts.Utils;
using System;

namespace Assets.Scripts.GamePlayScene.Mechanics.Items
{
	public class BalloonItem : SpriteBasedItem
	{
		public override bool CanExplodeDueToExplodeInNeigbour(MatchType sourceType)
		{
			return true;
		}

		public override void BringToFront()
		{
			Sorting sorting = this.GetSorting();
			sorting.Order += 120;
			this.ChangeSorting(sorting);
		}

		public override void SendToBack()
		{
			this.OnCellChanged();
		}

		public override ItemType GetItemType()
		{
			return ItemType.Balloon;
		}

		public override int GetScore()
		{
			return 500;
		}

		public override bool CanCastShadow()
		{
			return true;
		}

		public override void ReactToUpperFall()
		{
			FallListener fallListener = base.GetFallListener();
			fallListener.EndFall(-1f);
		}

		public override void PlayExplodeAudio()
		{
			AudioManager.Play(AudioTag.ItemBalloon, PlayMode.Frame, null, 1f);
		}

		public override CollectAnimation GetCollectItem()
		{
			CollectAnimation collectItem = base.GetCollectItem();
			collectItem.get_transform().set_position(this.CurrentSpriteRenderer.get_transform().get_position());
			return collectItem;
		}

		protected override void PlayOnExplodeStartedAnimation()
		{
			if (ParticlePool.Instance != null)
			{
				BalloonParticlePlayer balloonParticlePlayer = ParticlePool.Instance.PlayParticle(ParticlePool.Instance.BalloonExplosionParticle, base.get_transform().get_position(), true) as BalloonParticlePlayer;
				balloonParticlePlayer.SetColor();
			}
		}
	}
}

File: classes/game/item/BalloonItem.cpp
#include "BalloonItem.h"
#include "../Animation.h"
#include "../ParticleLibrary.h"
#include "../ImageLibrary.h"
#include "global/SoundManager.h"


BalloonItem::BalloonItem() {
}


BalloonItem::~BalloonItem() {
}

void BalloonItem::SetupItem() {
	ImageLibrary::CopySprite(this->CurrentSpriteRenderer, IMGLIB->Balloon);
	cocos2d::Rect newrect = CurrentSpriteRenderer->getTextureRect();
	newrect.origin.y -= 22.0f; newrect.size.height += 22.0f;
	CurrentSpriteRenderer->setTextureRect(newrect);
}
void BalloonItem::RemoveSelf() {
	this->OnBeforeRemove();
	//base.get_gameObject().Recycle();
	this->CurrentSpriteRenderer->removeFromParentAndCleanup(true);
	//delete this;
}

void BalloonItem::BringToFront() {
}

void BalloonItem::ReactToUpperFall() {
	//SpriteBasedItem::GetFallListener()->EndFall(-1.0f);
}

void BalloonItem::PlayExplodeAudio() {
	//SoundManager::playEffectSound(SoundManager::SoundEffect::S6); //AudioManager::getInstance()->Play(AudioTag.ItemBalloon, PlayMode.Frame, null, 1f);
}

CollectAnimation BalloonItem::GetCollectItem() {
	SpriteBasedItem::GetCollectItem();
	//CollectAnimation collectItem = Item::GetCollectItem();
	//collectItem->setPosition(this->CurrentSpriteRenderer->getPosition());
	return CollectAnimation::Instant; //collectItem;
}

void BalloonItem::PlayOnExplodeStartedAnimation() {
	PTCLLIB->explodeBalloon(this->CurrentSpriteRenderer->getPosition()); /*ORG:
	if (ParticlePool.Instance != null) {
		BalloonParticlePlayer balloonParticlePlayer = ParticlePool.Instance.PlayParticle(ParticlePool.Instance.BalloonExplosionParticle, base.get_transform().get_position(), true) as BalloonParticlePlayer;
		balloonParticlePlayer.SetColor();
	}*/
}